Override colour transform

AlanRace opened the pull request:
Added the ability to set the colour transform applied (resolves https://github.com/image-rs/jpeg-decoder/issues/238).

To do this, I had added a new `enum` which captures the possible options, including a `Default` option which retains the previous behaviour when the `color_transform == None`.

HeroicKatora commented on src/decoder.rs:
```diff
@@ -74,6 +74,31 @@ pub struct ImageInfo {
 
+/// Describes the colour transform to apply before binary data is returned
+#[derive(Debug, Clone, Copy)]
+pub enum ColorTransform {
```
Please add `#[non_exhaustive]` on here, just in case.

HeroicKatora commented on src/decoder.rs:
```diff
@@ -492,7 +520,8 @@ impl<R: Read> Decoder<R> {
                         match data {
                             AppData::Adobe(color_transform) => {
-                                self.color_transform = Some(color_transform)
+                                self.color_transform =
+                                    ColorTransform::AdobeColorTransform(color_transform)
```
This seems like it leads to confusing semantics. If one explicitly sets up `ColorTransform::None` but then it's overwritten by an app segment? Which semantics did you intend, could you document them on `set_transform` as well?

AlanRace replied:
Hmm this is a good point. What do you think to:

```rust
/// Sets the colour transform to be applied before returning the data. If must_use is set to true, 
/// then this transform will be used for decoding, and app segments relating to colour transforms 
/// will be ignored. If must_use is set to false, then this transform will be used only in the case 
/// where no app segment overwrites it.
    pub fn set_color_transform(&mut self, transform: ColorTransform, must_use: bool)
```

HeroicKatora replied:
Not sure, I'm not too fond of `bool` parameters but would maybe use a separate function for 'hints' vs. overrides, for instance because not all variants may be applicable as hints. Also, a hint might have the semantics of falling back to the default behavior in case `CMYK` is not applicable where an override errors in 3-component color.

Choices, so many choices :smile:

AlanRace replied:
Not quite sure if I understand what you mean with 'hints', but we could do:

```rust
/// Colour transform to use if one is not set using `set_color_transform` or found in app segments.
pub fn set_fallback_color_transform(&mut self, transform: ColorTransform) {
...
}

/// Colour transform to use when decoding the image. App segments relating to colour transforms 
/// will be ignored.
pub fn set_color_transform(&mut self, transform: ColorTransform) {
...
}
```

Alternatively, we could split out the fallback based on number of channels, like so:

```rust
/// Colour transform to use when the data has 3 channels and if one is not set using `set_color_transform` or found in app segments.
pub fn set_fallback_color_transform_3_channels(&mut self, transform: ColorTransform) {
...
}

```rust
/// Colour transform to use when the data has 4 channels and if one is not set using `set_color_transform` or found in app segments.
pub fn set_fallback_color_transform_4_channels(&mut self, transform: ColorTransform) {
...
}

```

But those are some long function names....

HeroicKatora replied:
Option 1 actually looks good.

fintelia replied:
Yeah. Perhaps even `set_fallback_color_transform` vs. `override_color_transform`.

AlanRace commented:
Hopefully this should now resolve the previous ambiguity.

I decided that the presence of app segments should set `color_transform` rather than overwriting `fallback_color_transform`, but that's a simple change if you prefer it the other way.

AlanRace commented:
As this pull request is dealing with colour transforms, I had a look at solving https://github.com/image-rs/jpeg-decoder/issues/249 too. I have rewritten the logic for selecting the colour transform based on the information [here](https://entropymine.wordpress.com/2018/10/22/how-is-a-jpeg-images-color-type-determined/).

That (hopefully) removes the need for the fallback transform being set, so I have removed it. Let me know what you think.
